agentcfg/model.go

```diff
@@ -18,32 +18,51 @@
 package agentcfg
 
 import (
+	"crypto/md5"
 	"encoding/json"
 	"fmt"
-	"strings"
-)
-
-const (
-	// ServiceName keyword
-	ServiceName = "service.name"
-	// ServiceEnv keyword
-	ServiceEnv = "service.environment"
 )
 
 // Doc represents an elasticsearch document
 type Doc struct {
-	ID     string `json:"_id"`
-	Source Source `json:"_source"`
-}
-
-// Source represents the elasticsearch _source field of a document
-type Source struct {
-	Settings Settings `json:"settings"`
+	Settings Settings
+	ID       string
 }
 
 // Settings hold agent configuration
 type Settings map[string]string
 
+// NewDoc unmarshals given byte slice into t Doc instance
+func NewDoc(in []byte) (*Doc, error) {
+	var doc Doc
+	if len(in) == 0 {
+		return &doc, nil
+	}
+	var tmp tmpDoc
+	if err := json.Unmarshal(in, &tmp); err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(tmp.Source.Settings, &doc.Settings); err != nil {
+		return nil, err
+	}
+	doc.ID = tmp.hash()
+	return &doc, nil
+}
+
+type tmpDoc struct {
+	ID     string `json:"_id"`
+	Source struct {
+		Settings json.RawMessage `json:"settings"`
+	} `json:"_source"`
+}
+
+func (t *tmpDoc) hash() string {
+	h := md5.New()
+	h.Write([]byte(t.ID))
```

Comment:
why include the ID in the hash?

Reply:
So it is unique across different services. This is not necessary for the workflow we have, but it feels weird that services with the same config result in the same Etag. Adding the ID is not a concerning overhead. At the moment there is no `version` field or `timestamp` field that changes when the config is updated, so this seemed the best solution.

Reply:
I'm more worried about the reverse, deleting and re-adding the same settings for a service would result in two different etags for the same configuration.

Reply:
I don't think that would be a big issue, but I can change it to not consider the ID.